R/getExperimentCrossCorrelation.R

```diff
@@ -0,0 +1,577 @@
+    # If there are no numeric values, give an error
+    if( all(!inds) ){
+      stop("Assay, specified by 'abund_values', of 'experiment1' does not include",
+           " numeric values. Choose categorical method for 'method'.",
+           call. = FALSE)
+    }
+    else if (any(!inds)) {
+      warning("Considering only numeric annotations for \n       
+                    pearson/spearman")
+    }
+  } else if (method %in% categorical_methods) {
+    inds <- vapply(assay, is.factor, TRUE)
+    # If there are no factor values, give an error
+    if( all(!inds) ){
+      stop("Assay, specified by 'abund_values', of 'experiment1' does not include",
+           " factor values. Choose numeric method for 'method'.",
+           call. = FALSE)
+    }
+    else if (any(!inds)) {
+      warning("Considering only categorical annotations for factors")
+    }
+  }
+  # Names of features that are numeric or factor (based on previous step)
+  names <- names(which(inds))
+  # Subset assay to get only numeric or factor values
+  if (!is.vector(assay)) {
+    assay <- suppressWarnings(as.matrix(assay[, inds], ncol=length(inds)))
+  } else {
+    assay <- as.matrix(assay[inds], ncol=length(inds))
+  }
+  # Add colnames to assay
+  colnames(assay) <- names
+  return(assay)
+}
+
+############################# .calculate_correlation ###########################
+.calculate_correlation <- function(assay1, assay2, method, p_adj_method){
+  # Create empty matrices
+  correlations <- matrix(NA, ncol(assay1), ncol(assay2))
+  rownames(correlations) <- colnames(assay1)
+  colnames(correlations) <- colnames(assay2)
+  p_values <- correlations
+  
+  # Calculate correlations, different methods for numeric and categorical data
+  if (method %in% c("pearson", "spearman")) {
+    # Specify minimum number of observation needed to do correlation analysis
+    minimum_number_of_observations <- 8
+    # Loop over every feature in assay2
+    for (j in seq_len(ncol(assay2))) {
+      # Loop over every feature in assay1
+      jc <- apply(assay1, 2, function(xi) {
+        # If there are enough observations/samples, do correlation test, and store the result
+        # to temporary object
+        if (sum(!is.na(xi)) >= minimum_number_of_observations) {
+          res <- suppressWarnings(
+            cor.test(xi, unlist(assay2[, j], use.names=FALSE), 
+                     method=method, use="pairwise.complete.obs"))
+          res <- c(res$estimate, res$p.value)
+        } 
+        # If there are not enough observations/samples, give a warning, and store
+        # just NAs
+        else {
+          warning(paste("Not enough observations (",
+                        minimum_number_of_observations, "required); \n   
+                        (", 
+                        sum(!is.na(xi)), ") \n \n 
+                        - skipping correlation estimation"))
+          res <- c(NA, NA)
+        }
+        # Return res which will be stored in jc object
+        res
+      })
```

Comment:
Now you can simplify this even futher like this:

```r
if(non-categorical){
  FUN1 <- function()
  FUN2 <- function()
} else {
  FUN1 <- function()
  FUN2 <- function()
}
apply(x,2,FUN1, FUN_internal = FUN2)
```

In this case FUN1 has to support the FUN_internal argument.

In addition you can try, not splitting the matrix with apply, but iterating over the matrices with indices, like this:

```r
i <- seq_len(ncol(assay1))
j <- seq_len(ncol(assay2))
lapply(i,function(){
lapply(j,function(){
})
})
``` 

Since subsetting is faster in R, than splitting the object, this approach can be faster some situations. The only way to find out is benchmarking.